### DISCUSSION TITLE
ContainerCreate shell can't run correct

### DISCUSSION THREAD

**Discussion by leon0204** [moby/moby#44787] [Q&A]
### Description

```go
EnvStr := []string{
"MYSQL_ROOT_PASSWORD=291838323"
}

resp, err := cli.ContainerCreate(ctx, &container.Config{
Image: imageName,
Env: EnvStr,
Shell: []string{"bash", "-c", "echo slow_query_log=on", ">>", "/etc/mysql/conf.d/mysql.conf.cnf"},
},nil,nil, nil, "irisMysql2")
```

```bash
docker container inspect irisMysql2

"Env": [
                "MYSQL_ROOT_PASSWORD=291838323",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "GOSU_VERSION=1.14",
                "MYSQL_MAJOR=8.0",
                "MYSQL_VERSION=8.0.31-1.el8",
                "MYSQL_SHELL_VERSION=8.0.31-1.el8"
            ],
            "Cmd": [
                "mysqld"
            ],
            "Image": "mysql:8.0",
            "Volumes": {
                "/var/lib/mysql": {}
            },
            "WorkingDir": "",
            "Entrypoint": [
                "docker-entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {},
            "Shell": [
                "bash",
                "-c",
                "echo slow_query_log=on",
                ">>",
                "/etc/mysql/conf.d/mysql.conf.cnf"
            ]
        },
```

### Reproduce

1 .  create the container
2 . docker exec into the container
3 . shell not run  correct

### Expected behavior

mysql-container- shell run 

### docker version

```bash
docker version
Client: Docker Engine - Community
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 18:04:24 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       3056208
  Built:            Tue Oct 25 18:02:38 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.9
  GitCommit:        1c90a442489720eec95342e1789ee8a5e1b9536f
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
[root@leon-tty ~]#
```

### docker info

```bash
[root@leon-tty ~]# docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  scan: Docker Scan (Docker Inc., v0.21.0)

Server:
 Containers: 136
  Running: 71
  Paused: 0
  Stopped: 65
 Images: 80
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1c90a442489720eec95342e1789ee8a5e1b9536f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1160.76.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.51GiB
 Name: leon-tty
 ID: 34KG:VFAH:6YP5:WFUX:57HE:5SMZ:EHQG:S2EX:ONPC:247A:7R7T:EDHW
 Docker Root Dir: /mnt/sdc/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  172.30.0.0/16
  127.0.0.0/8
 Registry Mirrors:
  https://hub-mirror.c.163.com/
  https://docker.mirrors.ustc.edu.cn/
 Live Restore Enabled: false
```

### Additional Info

golang version 
go version go1.18 darwin/amd64

import "github.com/docker/docker/client"

**Comment by thaJeztah** [comment]
It looks like you're trying to start the container, and to pass custom configuration options before the container is started (using a command run in `bash`). The `Shell` option is meant to configure a custom shell to be used as default, and not to run other commands.

A container expects a single command to be run as main process for the container; for the MySQL image, that command defaults to `mysqld`, but it uses an entrypoint script to prepare the container before running; https://github.com/docker-library/mysql/blob/e0d43b2a29867c5b7d5c01a8fea30a086861df2b/8.0/Dockerfile.oracle#L120-L123

While it's possible to override those commands; the `docker-entrypoint.sh` entrypoint script is _very_ likely still needed to run the container succesfully (as it does various preparation steps). So some options instead are;

- use a customized version of the entrypoint script (but this requires duplicating the default script, and keeping it in sync)
- configure the container with a custom configuration for mysql (this can be either by extending the image, or bind-mounting the custom configuration file(s)). For more info on that see the "Using a custom MySQL configuration file" section for the MySQL image: (https://hub.docker.com/_/mysql)

**Reply by leon0204** [reply]
The situation is that I am using golang to create mysql container. I want to create mysql.cnf configuration file when I create it, but the business situation does not allow me to directly mount the file, so I want to find a way to write the configuration file.

Therefore, I tried to write the configuration file before mysql started by using shell, Entrypoint, and other methods according to the api, but it was useless.

What I want to know is how to correctly use the api to create configuration files through echo without affecting mysqld running。https://github.com/moby/moby/blob/master/api/types/container/config.go#:~:text=type%20Config%20struct,Footer
